subprojects/core/src/main/java/org/gradle/api/internal/cache/DefaultCacheConfigurations.java

```diff
@@ -214,35 +224,44 @@ private static <T> Provider<T> providerFromSupplier(Supplier<T> supplier) {
 
     static abstract class DefaultCacheResourceConfiguration implements CacheResourceConfigurationInternal {
         private final String name;
-        private final Property<Long> removeUnusedEntriesOlderThan;
+        private final Clock clock;
+        private final Property<Long> entryRetentionAge;
+        private final Property<Long> entryRetentionTimestamp;
 
         @Inject
-        public DefaultCacheResourceConfiguration(PropertyHost propertyHost, String name) {
+        public DefaultCacheResourceConfiguration(PropertyHost propertyHost, String name, Clock clock) {
             this.name = name;
-            this.removeUnusedEntriesOlderThan = new ContextualErrorMessageProperty<>(propertyHost, Long.class, "removeUnusedEntriesOlderThan");
+            this.clock = clock;
+            this.entryRetentionAge = new ContextualErrorMessageProperty<>(propertyHost, Long.class, "entryRetentionAge");
+            this.entryRetentionTimestamp = new ContextualErrorMessageProperty<>(propertyHost, Long.class, "entryRetentionTimestamp");
         }
 
         @Override
-        public Property<Long> getRemoveUnusedEntriesOlderThan() {
-            return removeUnusedEntriesOlderThan;
+        public Property<Long> getEntryRetentionAge() {
+            return entryRetentionAge;
         }
 
-        /**
-         * @implNote Returns a supplier mapped from the property.  This provides a supplier that is resilient
-         * to subsequent changes to the property value as opposed to just calling get() on the property.
-         */
         @Override
-        public Supplier<Long> getRemoveUnusedEntriesOlderThanAsSupplier() {
-            return () -> getRemoveUnusedEntriesOlderThan().get();
+        public Property<Long> getEntryRetentionTimestamp() {
+            return entryRetentionTimestamp;
         }
 
+        @Override
+        public Supplier<Long> getEntryRetentionTimestampSupplier() {
+            return () -> {
+                if (entryRetentionTimestamp.isPresent()) {
+                    return entryRetentionTimestamp.get();
+                }
+                return clock.getCurrentTime() - entryRetentionAge.get();
```

Comment:
🤔 In our discussion, we talked about using the build start time here instead of the current time, to make the retention timestamp consistent across caches. Not sure if that is important or not. I suppose it doesn't matter too much, right?

Reply:
I tried to do this, but there was no available service that could be used to inject the build start time into the `CacheConfigurations` service. This makes sense since the cache-cleanup logic lives in the Gradle Daemon and is registered in `GradleUserHomeScopeServices`.

The trigger for cache cleanup isn't necessarily a build starting. It can also run on daemon shutdown. So I think using the current time works: this is no change from the previous behaviour.